### ISSUE TITLE
ZMQ hangs when it calls ZmqContext.Terminate

### ISSUE THREAD

**Issue by pray1997** [zeromq/clrzmq#129]
```
            using (ZeroMQ.ZmqContext ctx = ZeroMQ.ZmqContext.Create())
            using (ZeroMQ.ZmqSocket client = ctx.CreateSocket(SocketType.PUSH))
            {
                client.Connect("tcp://172.16.155.170:5566");
                byte[] title = Encoding.UTF8.GetBytes("Title");
                client.Send(title, title.Length, SocketFlags.None);
                client.Close();
                ctx.Terminate();
            }
```

**Comment by joshualangemann** [comment]
Should you be closing the socket before terminating the context? Try:

using (ZeroMQ.ZmqContext ctx = ZeroMQ.ZmqContext.Create())
{
    using (ZeroMQ.ZmqSocket client = ctx.CreateSocket(SocketType.PUSH))
    {
        client.Connect("tcp://172.16.155.170:5566");
        byte[] title = Encoding.UTF8.GetBytes("Title");
        client.Send(title, title.Length, SocketFlags.None);
        client.Close();
    }
    ctx.Terminate();
}
- Josh

On Mar 27, 2013, at 7:57 AM, pray1997 <notifications@github.com<mailto:notifications@github.com>>
 wrote:

```
        using (ZeroMQ.ZmqContext ctx = ZeroMQ.ZmqContext.Create())
        using (ZeroMQ.ZmqSocket client = ctx.CreateSocket(SocketType.PUSH))
        {
            client.Connect("tcp://172.16.155.170:5566");
            byte[] title = Encoding.UTF8.GetBytes("Title");
            client.Send(title, title.Length, SocketFlags.None);
            client.Close();
            ctx.Terminate();
        }
```

**Comment by pray1997** [comment]
it's same, you can try it

**Comment by joshualangemann** [comment]
Since you are using the "using" construct, you shouldn't have to explicitly close the socket or terminate the context. Using should do that for you. You would only need to call "client.Close()" or "ctx.Terminate()" explicitly if you were not doing it already via "using".

**Comment by pray1997** [comment]
yes, you are right, i don't need to terminate the context explicitly. 
But if you try the code I posted above, you will find that the application will hang at the end "}" of using.
And when I digged into the source code, I found it's still due to Terminate method.

**Comment by TimGunnells** [comment]
If you add:

```
                client.Linger = System.TimeSpan.FromSeconds(1);
```

it should stop hanging on the terminate

**Comment by pray1997** [comment]
Yes! It works! Thanks a lot!

```
     client.Linger = System.TimeSpan.FromSeconds(1);
```

**Comment by Robert--** [comment]
Here is a C# program that hangs for me (regardless of linger). I know calling client.Close() before the context.Terminate() will let the program exit. But there is a bunch of underlying libzmq logic in place, during the Terminate(), to signal a slot for the existing socket, expecting the reaper thread to wake up and do something which it never does. Looks like there is some bug here. 

<pre><code>
            ZmqContext context = ZmqContext.Create();
            var client = context.CreateSocket(SocketType.REQ);
            //client.Linger = System.TimeSpan.FromSeconds(1);
            context.Terminate();
</code></pre>
